### PR TITLE
[mle] fix the issue when the number of registered addresses exceeds the parent's capacity

### PR DESCRIPTION
**Pull request by librasungirl**
OT would fail to process address registration with more than 4 entries as `IsValid()` would always fail.

This commit processes the address registration entry one by one directly from message.

### CODE REVIEW

**Comment by librasungirl** [comment]
Rebase

File: src/core/thread/mle_router.cpp [hunk 1]
```diff
@@ -2081,46 +2081,59 @@ otError MleRouter::SendParentResponse(Child *aChild, const ChallengeTlv &aChalle
...
+    uint8_t registeredCount = 0;
+    uint8_t storedCount = 0;
+    uint16_t offset = aOffset;
+    uint16_t end = 0;
     char stringBuffer[Ip6::Address::kIp6AddressStringSize];
 
+    VerifyOrExit(aMessage.Read(offset, sizeof(tlv), &tlv) == sizeof(tlv));
+
+    offset += sizeof(tlv);
+    end = offset + tlv.GetLength();
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: Validate that the TLV fits within the message (i.e. `tlv.GetLength() <= aMessage.GetLength() - aOffset`)?

File: src/core/thread/mle_router.cpp [hunk 2]
```diff
@@ -2269,6 +2275,12 @@ otError MleRouter::HandleChildIdRequest(const Message &aMessage, const Ip6::Mess
         VerifyOrExit(pendingTimestamp.IsValid(), error = OT_ERROR_PARSE);
     }
 
+    if ((mode.GetMode() & ModeTlv::kModeFFD) == 0)
+    {
+        SuccessOrExit(error = Tlv::GetOffset(aMessage, Tlv::kAddressRegistration, addressRegistrationOffset));
+        UpdateChildAddresses(aMessage, addressRegistrationOffset, *child);
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: Handle parse errors?

File: src/core/thread/mle_router.cpp [hunk 3]
```diff
@@ -2408,10 +2418,9 @@ otError MleRouter::HandleChildUpdateRequest(const Message &aMessage, const Ip6::
     }
 
     // Ip6 Address TLV
-    if (Tlv::GetTlv(aMessage, Tlv::kAddressRegistration, sizeof(address), address) == OT_ERROR_NONE)
+    if (Tlv::GetOffset(aMessage, Tlv::kAddressRegistration, addressRegistrationOffset) == OT_ERROR_NONE)
     {
-        VerifyOrExit(address.IsValid(), error = OT_ERROR_PARSE);
-        UpdateChildAddresses(address, *child);
+        UpdateChildAddresses(aMessage, addressRegistrationOffset, *child);
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: Handle parse errors?

File: src/core/thread/mle_router.cpp [hunk 4]
```diff
@@ -2546,10 +2555,9 @@ otError MleRouter::HandleChildUpdateResponse(const Message &aMessage, const Ip6:
     }
 
     // Ip6 Address
-    if (Tlv::GetTlv(aMessage, Tlv::kAddressRegistration, sizeof(address), address) == OT_ERROR_NONE)
+    if (Tlv::GetOffset(aMessage, Tlv::kAddressRegistration, addressRegistrationOffset) == OT_ERROR_NONE)
     {
-        VerifyOrExit(address.IsValid(), error = OT_ERROR_PARSE);
-        UpdateChildAddresses(address, *child);
+        UpdateChildAddresses(aMessage, addressRegistrationOffset, *child);
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: Handle parse errors?

File: src/core/thread/mle_router.cpp (hunk 1 above)
Comment (librasungirl) [reply src/core/thread/mle_router.cpp]: Updated. Thanks.

File: src/core/thread/mle_router.cpp (hunk 2 above)
Comment (librasungirl) [reply src/core/thread/mle_router.cpp]: Updated. Thanks.

**Pull request merged**